metadata-ingestion/src/datahub/sdk/lineage_client.py

```diff
@@ -227,9 +240,140 @@ def add_dataset_transform_lineage(
         if query_entity:
             mcps.extend(query_entity)
         self._client._graph.emit_mcps(mcps)
+
+    def add_dataset_lineage_from_sql(
+        self,
+        *,
+        query_text: str,
+        platform: str,
+        platform_instance: Optional[str] = None,
+        env: str = "PROD",
+        default_db: Optional[str] = None,
+        default_schema: Optional[str] = None,
+        schema_aware: bool = True,
+    ) -> None:
+        """Add lineage by parsing a SQL query."""
+        from datahub.sql_parsing.sqlglot_lineage import create_lineage_sql_parsed_result
+
+        # Parse the SQL query to extract lineage information
+        parsed_result = create_lineage_sql_parsed_result(
+            query=query_text,
+            default_db=default_db or "",
+            default_schema=default_schema or "",
+            platform=platform,
+            platform_instance=platform_instance,
+            env=env,
+            graph=self._client._graph,
+            schema_aware=schema_aware,
+        )
+
+        # Validate parsing result
+        if not parsed_result:
+            raise SdkUsageError("Failed to parse SQL query: Unable to parse")
+        if parsed_result.debug_info.error:
+            raise SdkUsageError(
+                f"Failed to parse SQL query: {parsed_result.debug_info.error}"
+            )
+        if not parsed_result.out_tables:
+            raise SdkUsageError(
+                "No output tables found in the query. Cannot establish lineage."
+            )
+
+        # Use the first output table as the downstream
+        downstream_urn = parsed_result.out_tables[0]
+
+        # Process all upstream tables found in the query
+        for i, upstream_table in enumerate(parsed_result.in_tables):
+            # Skip self-lineage
+            if upstream_table == downstream_urn:
+                continue
+
+            # Extract column-level lineage for this specific upstream table
+            column_mapping = {}
+            if parsed_result.column_lineage:
+                for col_lineage in parsed_result.column_lineage:
+                    if not (col_lineage.downstream and col_lineage.downstream.column):
+                        continue
+
+                    # Filter upstreams to only include columns from current upstream table
+                    upstream_cols = [
+                        ref.column
+                        for ref in col_lineage.upstreams
+                        if ref.table == upstream_table and ref.column
+                    ]
+
+                    if upstream_cols:
+                        column_mapping[col_lineage.downstream.column] = upstream_cols
+
+            # Add lineage, including query text only for the first upstream
+            self.add_dataset_transform_lineage(
+                upstream=upstream_table,
+                downstream=downstream_urn,
+                column_lineage=column_mapping or None,
+                query_text=query_text if i == 0 else None,
```

Comment:
this doesn't look right - all the emits should be done in one go, and the query urn should be attached to all of the lineage

Reply:
updated query_text to be attached to all of the lineage 

this creates table-level lienage with shared query enitites + CLL. is this something that we expect?
![Screenshot 2025-05-07 at 9 33 20 PM](https://github.com/user-attachments/assets/c658e8f9-a8f7-4b03-a8ff-fa26b9eb158c)

Reply:
> all the emits should be done in one go

I'm not sure I follow this -- isn't it simpler to do `add_dataset_transform_lineage` for each upstream table while iterating it? the alternative that I can think of is use `datasetPatchbuilder.set_upstream_lineages` and `set_fine_grained_upstream_lineages`, and I'm not seeing any specific benefit with it

lmk if I miss anything here!

Reply:
the query shoudl be attached to CLL as well 
should've workd from this -- 

https://github.com/datahub-project/datahub/blob/72aab9fe636681c89ce2e69c8dc200e779c5b489/metadata-ingestion/src/datahub/sdk/lineage_client.py#L220


golden file shoudl look like this https://github.com/datahub-project/datahub/blob/a4bce6af1c0a20cb24b8e86a717f601b4d2f31c8/metadata-ingestion/tests/unit/sql_parsing/aggregator_goldens/test_basic_lineage.json#L0-L1